pkg/remote/remote.go

```diff
@@ -166,13 +216,186 @@ func (c Client) DownloadFile(localFile string) error {
+	file, err := os.Open(filename)
+	if err != nil {
+		log.Warnf("failed to read known_hosts files: %s", err)
+		return hostKeys
+	}
+
+	defer file.Close()
+
+	scanner := bufio.NewScanner(file)
+
+	for scanner.Scan() {
+		fields := strings.Split(scanner.Text(), " ")
+		if len(fields) < 3 {
+			continue
+		}
+
+		hostnames := strings.Split(fields[0], ",")
+
+		hostSet := HostSet{}
+
+		for _, host := range hostnames {
+			if host == "" {
+				continue
+			}
+
+			hostSet[strings.ToLower(host)] = struct{}{}
+		}
+
+		if hostSet.Contains([]string{c.HostKeyAlias, c.HostDisplayName, c.HostDerived}) {
+			hostKey, _, _, _, err := ssh.ParseAuthorizedKey(scanner.Bytes())
+			if err != nil {
+				log.Warnf("failed to parse %q: %s", fields[2], err)
+			} else {
+				hostKeys = append(hostKeys, hostKey)
+			}
+		}
+	}
+
+	return hostKeys
+}
+
+func (c Client) strictHostKeyChecking() string {
+	strict := ssh_config.Get(c.HostDisplayName, "StrictHostKeyChecking")
+
+	if strict == "" && c.HostDisplayName != c.HostDerived {
+		strict = ssh_config.Get(c.HostDerived, "StrictHostKeyChecking")
+	}
+
+	if strict == "" {
+		strict = ssh_config.Default("StrictHostKeyChecking")
+	}
+
+	if strict == "accept-new" || strict == "off" {
+		strict = "no"
+	}
+
+	return strict
+}
+
+// getKnownHostsFile returns the path to the known hosts file.
+func (c Client) getKnownHostsFile() (string, error) {
+	file := ssh_config.Get(c.HostDisplayName, "UserKnownHostsFile")
+
+	if file == "" && c.HostDisplayName != c.HostDerived {
+		file = ssh_config.Get(c.HostDerived, "UserKnownHostsFile")
+	}
+
+	if file == "" {
+		file = "~/.ssh/known_hosts"
+	}
+
+	return homedir.Expand(file)
+}
+
+// getIdentityFile returns the path to a secret key file, or the first existing default.
+func (c Client) getIdentityFile() string {
+	keyFile := ssh_config.Get(c.HostDisplayName, "IdentityFile")
+
+	if keyFile == "" && c.HostDisplayName != c.HostDerived {
+		keyFile = ssh_config.Get(c.HostDerived, "IdentityFile")
+	}
+
+	// check for defaults on disk
+	if keyFile == "" {
+		for _, key := range defaultKeyFiles {
+			keyExpanded, err := homedir.Expand(key)
+			if err != nil {
+				continue
+			}
+
+			if _, err := os.Stat(keyExpanded); os.IsNotExist(err) {
+				continue
+			}
+
+			keyFile = key
+
+			break
+		}
+	}
+
+	if keyFile == "" {
+		return ""
+	}
+
+	keyFile, _ = homedir.Expand(keyFile)
+
+	return keyFile
+}
+
+func (c Client) signerForIdentity() ssh.Signer {
+	identityFile := c.getIdentityFile()
+	if identityFile == "" {
+		return nil
+	}
+
+	key, err := ioutil.ReadFile(identityFile)
 	if err != nil {
-		log.Errorf("failed to get host keys: %s", err)
+		log.Warnf("unable to read private key %s: %v", identityFile, err)
+		return nil
 	}
 
+	signer, err := ssh.ParsePrivateKey(key)
+	if err != nil {
+		log.Fatalf("unable to parse private key %s: %v", identityFile, err)
+	}
+
+	return signer
+}
+
+func (c Client) warnIfUsingRevokedHostKeys() {
```

Comment:
It seems we can do a small refactoring and remove this method. Let's discuss later.